File: QuotaPetsShared/Sources/QuotaPetsShared/Usage/ProviderError.swift
import Foundation

/// Provider error states (§25).
///
/// `providerResponseChanged` is the important one: both usage endpoints are private and
/// unversioned (research §3.1), so a schema change is an expected operating condition,
/// not a crash. It is raised whenever a payload parses as JSON but does not carry the
/// fields we require.
public enum ProviderError: Error, Sendable, Hashable {
    case notAuthenticated
    case tokenExpired
    case networkUnavailable
    /// Retry hint from the server (`Retry-After`) when it supplied one.
    case rateLimited(retryAfter: TimeInterval?)
    case providerResponseChanged(detail: String)
    case serverError(status: Int)
    case unknown(detail: String)

    /// Compact state for the Watch. The brief (§25) forbids dumping raw HTTP onto the
    /// wrist; the phone's debug view gets the full case instead.
    public var watchLabel: String {
        switch self {
        case .notAuthenticated, .tokenExpired: "AUTH"
        case .networkUnavailable: "OFFLINE"
        case .rateLimited: "STALE"
        case .providerResponseChanged: "PROVIDER ERROR"
        case .serverError, .unknown: "PROVIDER ERROR"
        }
    }

    /// Whether a bounded retry is worth attempting (§26). Auth and schema failures are
    /// not retryable — retrying them just burns quota and battery to fail identically.
    public var isTransient: Bool {
        switch self {
        case .networkUnavailable, .serverError: true
        case .rateLimited: false
        case .notAuthenticated, .tokenExpired, .providerResponseChanged, .unknown: false
        }
    }
}

/// Maps an HTTP status to the error model. Mirrors Orca's classification
/// (research §2/§3) without importing its Electron-specific machinery.
public enum HTTPStatusClassifier {
    public static func error(for status: Int, retryAfter: TimeInterval? = nil) -> ProviderError? {
        switch status {
        case 200...299: nil
        case 401: .notAuthenticated
        case 403: .tokenExpired
        case 429: .rateLimited(retryAfter: retryAfter)
        case 500...599: .serverError(status: status)
        default: .unknown(detail: "HTTP \(status)")
        }
    }
}

File: src/usage/providerError.ts
/**
 * Provider error taxonomy.
 *
 * `providerResponseChanged` is the important one: both usage endpoints are private and
 * unversioned, so a schema change is an expected operating condition, not a crash. It is
 * raised whenever a payload parses as JSON but does not carry the fields we require.
 */
export type ProviderErrorKind =
  | "notAuthenticated"
  | "tokenExpired"
  | "networkUnavailable"
  | "rateLimited"
  | "providerResponseChanged"
  | "serverError"
  | "unknown";

export class ProviderError extends Error {
  readonly kind: ProviderErrorKind;
  readonly status: number | null;
  /** Seconds, from the server's Retry-After, when it sent one. */
  readonly retryAfterSec: number | null;

  constructor(
    kind: ProviderErrorKind,
    message: string,
    opts: { status?: number | null; retryAfterSec?: number | null } = {},
  ) {
    super(message);
    this.name = "ProviderError";
    this.kind = kind;
    this.status = opts.status ?? null;
    this.retryAfterSec = opts.retryAfterSec ?? null;
  }

  /**
   * Whether a bounded retry is worth attempting. Auth and schema failures are not
   * retryable — retrying them burns battery and quota to fail identically. `rateLimited`
   * is deliberately NOT transient: retrying a 429 actively makes things worse.
   */
  get isTransient(): boolean {
    return this.kind === "networkUnavailable" || this.kind === "serverError";
  }
}

export function isProviderError(e: unknown): e is ProviderError {
  return e instanceof ProviderError;
}

/** Maps an HTTP status to the error model, or null when the response is a success. */
export function classifyStatus(
  status: number,
  retryAfterSec: number | null = null,
): ProviderError | null {
  if (status >= 200 && status <= 299) return null;
  if (status === 401) return new ProviderError("notAuthenticated", "Not authenticated", { status });
  if (status === 403) return new ProviderError("tokenExpired", "Token rejected", { status });
  if (status === 429)
    return new ProviderError("rateLimited", "Rate limited", { status, retryAfterSec });
  if (status >= 500 && status <= 599)
    return new ProviderError("serverError", `Server error ${status}`, { status });
  return new ProviderError("unknown", `HTTP ${status}`, { status });
}

/**
 * RFC 9110 allows Retry-After to be either delay-seconds or an HTTP-date. Ignoring the
 * date form silently loses the hint on servers that use it.
 */
export function parseRetryAfter(header: string | null | undefined, now: number): number | null {
  if (!header) return null;
  const trimmed = header.trim();
  if (trimmed === "") return null;

  const asNumber = Number(trimmed);
  if (Number.isFinite(asNumber) && asNumber >= 0) return asNumber;

  const asDate = Date.parse(trimmed);
  if (Number.isNaN(asDate)) return null;
  return Math.max(0, (asDate - now) / 1000);
}
